filemanager/gcsmanager.go

```diff
@@ -85,18 +101,31 @@ func (m *gcsManager) Upload(ctx context.Context, file *os.File, prefixes ...stri
 	ctx, cancel := context.WithTimeout(ctx, m.getTimeout())
 	defer cancel()
 
-	obj := client.Bucket(m.config.Bucket).Object(fileName)
-	w := obj.NewWriter(ctx)
-	if _, err := io.Copy(w, file); err != nil {
-		err = fmt.Errorf("copying file to GCS: %v", err)
-		if closeErr := w.Close(); closeErr != nil {
-			return UploadedFile{}, fmt.Errorf("closing writer: %q, while: %w", closeErr, err)
-		}
+	object := client.Bucket(m.config.Bucket).Object(fileName)
+	if m.config.doNotOverWrite {
+		object = object.If(storage.Conditions{DoesNotExist: true})
+	}
+	w := object.NewWriter(ctx)
 
-		return UploadedFile{}, err
+	if _, err := io.Copy(w, file); err != nil {
+		return UploadedFile{}, fmt.Errorf("copying file to writer: %v", err)
 	}
 
 	if err := w.Close(); err != nil {
+		switch e := err.(type) {
+		case *googleapi.Error:
+			if e.Code == http.StatusPreconditionFailed {
+				m.logger.Warnn(
+					"object already exists",
+					logger.NewStringField("object", fileName),
+				)
+				return UploadedFile{
+					Location:   "not uploaded, already exists",
+					ObjectName: fileName,
+				}, nil
```

Comment:
Didn't get how the client of this functionality is going to handle it in the case of `StatusPreconditionFailed` because they would expect some valid location but instead they will get `not uploaded, already exists`. Should we return a sentinel error to handle it?

Reply:
using `var ErrPreConditionFailed = errors.New("precondition failed")` now